main.lua

```diff
@@ -1366,3 +1350,71 @@ function userdataIterator(data)
         if success then return idx, item end
     end
 end
+
+local LSPCmds = {
+    ["start"] = startServer,
+    ["stop"] = stopServers,
+    ["showlog"] = showLog,
+    ["sync-document"] = function (bp) syncFullDocument(bp.Buf) end,
+    ["hover"] = hoverAction,
+    ["format"] = formatAction,
+    ["autocomplete"] = completionAction,
+    ["goto-definition"] = gotoAction("definition"),
+    ["goto-declaration"] = gotoAction("declaration"),
+    ["goto-typedefinition"] = gotoAction("typeDefinition"),
+    ["goto-implementation"] = gotoAction("implementation"),
+    ["find-references"] = findReferencesAction,
+    ["document-symbols"] = documentSymbolsAction,
+    ["diagnostic-info"] = openDiagnosticBufferAction,
+}
+
+local function lspCompleter(buf)
+    local opts = {}
+
+    --Do NOT autocomplete after first argument
+    local args = go_strings.Split(buf:Line(0), " ")
+    if #args > 2 then return nil, nil end
+
+    for k, _ in pairs(LSPCmds) do table.insert(opts, k) end
+    table.sort(opts)
+
+    --NOTE assumes no unicode in plugin's options
+    local suggestions = {}
+    local completions = {}
+    local lastArg = args[#args]
+
+    for i=1,#opts do
+        local opt = opts[i]
+        local optPrefix = string.sub(opt, 1, #lastArg)
+
+        if optPrefix > lastArg then break end
+
+        local startIdx, endIdx = string.find(opt, lastArg, 1, true)
+        if endIdx and startIdx == 1 then
```

Comment:
I think the first check here is redundant(?)
```suggestion
        if startIdx == 1 then
```

As a side note it might be nice to do completions like "lsp def" -> "lsp goto-definition"?

Reply:
If, for any reason, `endIdx` is `nil`, it will cause a runtime error in the next line. I don't recall if that actually occurred in my plugins or if it was just a safety measure.

Reply:
Either both `startIdx` and `endIdx` are `nil` or neither is. As far as I know `string.find` can't return nil for one but not the other.